Fix for issue #3008 Swift Pod causes runtime failure in Obj-C host app o...

jrosssavant opened the pull request:
Fix for issue #3008 Swift Pod causes runtime failure in Obj-C host app on device (codesign issue)

segiddins commented:
Please use the existing `code_sign` function and add tests that show this fixes things (likely by updating an example app or creating a new one)

jrosssavant commented:
@segiddins Can you suggest how to write a test that will confirm this fix?  This issue occurs only at runtime (not during the build), and only on a device (not in the simulator).  Is there an existing test that runs on the device only?  Sorry if this seems obvious and I'm missing it - I'm relatively new to CocoaPods and Ruby.

jrosssavant commented:
Cleaned up the fix a bit and added integration spec fixes here https://github.com/CocoaPods/cocoapods-integration-specs/pull/24.  I'll next take a look at a more specific test to verify this fix directly.

jrosssavant commented:
@segiddins I added an example app which demonstrates this issue.  Please let me know if you need any more information about this fix or pull request.  Thanks!

fluidsonic commented:
I don't understand why CocoaPods copies the Swift standard libraries in the first place which just slows down build times.
Xcode is copying & code signing and thus overwriting these libraries anyway. It's just that CocoaPods sometimes overwrites them **again** which causes random code signing errors.

Whether Xcode copies Swift standard libraries can be controlled using the [`EMBEDDED_CONTENT_CONTAINS_SWIFT`](https://developer.apple.com/library/ios/qa/qa1881/_index.html) build setting.

neonichu commented:
@fluidsonic unfortunately, that build settings does not work in all cases, see <http://samdmarshall.com/blog/swift_and_objc.html>

segiddins commented on lib/cocoapods/generator/embed_frameworks_script.rb:
```diff
@@ -78,6 +78,7 @@ def script
             for lib in $swift_runtime_libs; do
               echo "rsync -av \\"${SWIFT_STDLIB_PATH}/${lib}\\" \\"${destination}\\""
               rsync -av "${SWIFT_STDLIB_PATH}/${lib}" "${destination}"
+              code_sign "${destination}/${lib}"
```
This should check is code signing is enabled, like is done above.

jrosssavant replied:
Ah yes, thanks for catching that!

jrosssavant commented:
@segiddins I think all the feedback on this PR has been addressed.  Is there anything else you need from me to land this?  The Travis CI builds won't pass unless they can somehow find my updated integration spec submodule branch - maybe if that branch is pulled into the cocoa-integration-specs repo before merging?

segiddins commented:
Can @mrackwitz or @neonichu or @samdmarshall take a look at this one?

mrackwitz commented:
The change itself looks good. I'm not sure if we need another example for this, as this still doesn't really cover, what we want to test here. The rake task builds them for the iOS simulator, so codesigning is not required, and I don't see how we would test it for development and CI.

jrosssavant commented:
True.  The bug only manifests at runtime on a device.  Today I was testing on an iPhone 5 running iOS 7.1.2 device as well as my iPhone 6, and this bug did not occur on the iOS 7 device.  I'm guessing this has something to do with how embedded frameworks are supported for iOS 7.  Automated testing would require an actual device be attached (and would still be tricky even then.) 

 I also tried messing around with the `codesign --verify` command to see if I could verify this after the build, but I couldn't find a way for it to detect the problem.

Please let me know if you'd like for me to pull the example or leave it in there.

segiddins commented:
@jrosssavant please rebase + delete the example app and I'll merge

jrosssavant commented:
Will do, thanks.

jrosssavant commented:
Alright, I've rebased the commits for both CocoaPods and the integration specs, removing the example app.  Please double-check this and let me know if more changes are needed, or merge at your earliest convenience.
